drivers/adc/adc_gd32.c

```diff
@@ -0,0 +1,435 @@
+	/* Set SWRCST as the regular channel external trigger. */
+	ADC_CTL1(cfg->reg) &= ~ADC_CTL1_ETSRC;
+	ADC_CTL1(cfg->reg) |= CTL1_ETSRC(7);
+
+	/* Enable external trigger for regular channel. */
+	ADC_CTL1(cfg->reg) |= ADC_CTL1_ETERC;
+#endif
+
+	/* Enable ADC */
+	ADC_CTL1(cfg->reg) |= ADC_CTL1_ADCON;
+
+	adc_gd32_calibration(cfg);
+
+	cfg->irq_config_func();
+
+	adc_context_unlock_unconditionally(&data->ctx);
+
+	return 0;
+}
+
+#define HANDLE_SHARED_IRQ(n, active_irq)							\
+	static const struct device *dev_##n = DEVICE_DT_INST_GET(n);				\
+	const struct adc_gd32_config *cfg_##n = dev_##n->config;				\
+												\
+	if ((cfg_##n->irq_num == active_irq) &&							\
+		(ADC_CTL0(cfg_##n->reg) & ADC_CTL0_EOCIE)) {					\
+		adc_gd32_isr(dev_##n);								\
+	}
+
+static void adc_gd32_global_irq_handler(const struct device *dev)
+{
+	const struct adc_gd32_config *cfg = dev->config;
+
+	LOG_DBG("global irq handler: %u", cfg->irq_num);
+
+	DT_INST_FOREACH_STATUS_OKAY_VARGS(HANDLE_SHARED_IRQ, (cfg->irq_num));
+}
+
+#ifdef CONFIG_SOC_SERIES_GD32F3X0
+#define ADC_CLOCK_SOURCE(n)									\
+	.rcu_clock_source = DT_INST_PROP(n, rcu_periph_clock)
+#else
+#define ADC_CLOCK_SOURCE(n)
+#endif
+
+#define DT_ADC2 0U
+
+#define ADC_GD32_INIT(n)									\
+	PINCTRL_DT_INST_DEFINE(n);								\
+	static void adc_gd32_irq_cfg_##n(void)							\
+	{											\
+		bool initialized = false;							\
+												\
+		for (size_t i = 0; i < sizeof(adc_gd32_init_irqs); i++) {			\
+			if (adc_gd32_init_irqs[i] == 0U) {					\
+				initialized = true;						\
+				adc_gd32_init_irqs[i] = DT_INST_IRQN(n);			\
+				break;								\
+			}									\
+												\
+			if (adc_gd32_init_irqs[i] == DT_INST_IRQN(n)) {				\
+				LOG_INF("interrupt initialize success, num: %u",		\
+					DT_INST_IRQN(n));					\
+				return;								\
+			}									\
+		}										\
+												\
+		if (initialized) {								\
+			IRQ_CONNECT(DT_INST_IRQN(n),						\
+				DT_INST_IRQ(n, priority),					\
+				adc_gd32_global_irq_handler,					\
+				DEVICE_DT_INST_GET(n),						\
+				0);								\
+			irq_enable(DT_INST_IRQN(n));						\
+			LOG_INF("interrupt initialize success, num: %u, priority: %u",		\
+				DT_INST_IRQN(n), DT_INST_IRQ(n, priority));			\
+		} else {									\
+			LOG_INF("interrupt initialize failed");					\
+		}										\
+	}											\
+	static struct adc_gd32_data adc_gd32_data_##n = {					\
+		ADC_CONTEXT_INIT_TIMER(adc_gd32_data_##n, ctx),					\
+		ADC_CONTEXT_INIT_LOCK(adc_gd32_data_##n, ctx),					\
+		ADC_CONTEXT_INIT_SYNC(adc_gd32_data_##n, ctx),					\
+	};											\
+	const static struct adc_gd32_config adc_gd32_config_##n = {				\
+		.reg = DT_INST_REG_ADDR(n),							\
+		.rcu_periph_clock = DT_INST_PROP(n, rcu_periph_clock),				\
+		.channels = DT_INST_PROP(n, channels),						\
+		.pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(n),					\
+		.irq_num = DT_INST_IRQN(n),							\
+		.irq_config_func = adc_gd32_irq_cfg_##n,					\
+		ADC_CLOCK_SOURCE(n)								\
```

Comment:
why not:
```
IF_ENABLED(CONFIG_SOC_SERIES_GD32F3X0, (.rcu_clock_source = DT_INST_PROP(n, rcu_periph_clock),))
```

Reply:
There also have others gd32 soc contain additional adc clock.